typescript/token/contracts/FastHypERC20.sol

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.8.0;
+
+import {HypERC20} from "./HypERC20.sol";
+import {Message} from "./libs//Message.sol";
+
+import {ERC20Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+
+/**
+ * @title Hyperlane ERC20 Token Router that extends ERC20 with remote transfer functionality.
+ * @author Abacus Works
+ * @dev Supply on each chain is not constant but the aggregate supply across all chains is.
+ */
+contract FastHypERC20 is HypERC20 {
+    /**
+     * @notice `FastTransferMetadata` is the LP data stored against `fastTransferId`.
+     */
+    struct FastTranferMetadata {
+        address filler;
+        address recipient;
+        uint256 amount;
+        uint256 fastFee;
+    }
+
+    uint256 public fastTransferId;
+    // maps `fastTransferId` to metadata about the user who made the transfer.
+    mapping(uint256 => FastTranferMetadata) fastTransfers;
+
+    constructor(uint8 __decimals) HypERC20(__decimals) {}
+
+    /**
+     * @dev Mints `_amount` of token to `_recipient`/`fastFiller` who provided LP.
+     * @inheritdoc HypERC20
+     */
+    function _transferTo(
+        address _recipient,
+        uint256 _amount,
+        bytes calldata _metadata
+    ) internal override {
+        if (
+            _metadata.length > 0 &&
+            _fastTransferTo(_recipient, _amount, _metadata)
+        ) {
+            return; // Fast transfer succeeded, exit early
+        }
+
+        _mint(_recipient, _amount);
+    }
+
+    /**
+     * @dev `_fastTransferTo` allows the `_transferTo` function to send the token to the LP.
+     * @param _recipient The ricipiant of the token.
+     * @param _amount The amount of tokens that is bridged.
+     * @param _metadata Metadata is a byte array ofg (uint256, uint256).
+     */
+    function _fastTransferTo(
+        address _recipient,
+        uint256 _amount,
+        bytes calldata _metadata
+    ) internal returns (bool) {
+        (uint256 _fastFee, uint256 _fastTransferId) = abi.decode(
+            _metadata,
+            (uint256, uint256)
+        );
+
+        FastTranferMetadata memory m_filledMetadata = fastTransfers[
+            _fastTransferId
```

Comment:
Actually, a warp route can have multiple remotes, so you probably need to key this by the remote domain/router as well